File: 02_appendix_F/scripts/speed_noCI.jl
using Pkg
#Pkg.add(PackageSpec(url="https://github.com/flo1met/TargetTrialEmulation.jl", rev="dev"))
using TargetTrialEmulation
using CSV
using DataFrames


files = readdir("02_apendix_F/datasets/")
results = DataFrame(file=String[], time_seconds=Float64[], alloc_kb=Float64[])

for fname in files
    println("Processing $fname")
    df = CSV.read("02_apendix_F/datasets/$fname", DataFrame)

    timing = @timed begin
        df_out, model = TTE(df, 
            id_var = :ID,
            outcome = :Y, 
            treatment = :A, 
            period = :t, 
            eligible = :eligible, 
            censored = :C,
            ipcw = true,
            covariates = [:X2, :X4], 
            method = "ITT",
            save_w_model = false,
            fill_missing_timepoints = false,
            estimate_surv = false
        )
    end

    push!(results, (
        file = fname,
        time_seconds = timing.time,
        alloc_kb = timing.bytes / 1024
    ))
end

CSV.write("02_apendix_F/out/performance_summary_noCI_julia.csv", results)

results

File: 02_appendix_F/scripts/speed_noCI.R
library(TrialEmulation)
library(tidyverse)

files <- list.files("02_appendix_F/datasets/", full.names = TRUE)

results <- data.frame(file = character(), time_seconds = numeric(), alloc_kb = numeric())

for (f in files) {
  message("Processing ", f)
  dat <- read_csv(f, show_col_types = FALSE)

  # temporray profiling file
  mem_file <- tempfile(fileext = ".out")

  # start profiling memory
  Rprofmem(mem_file)

  t <- system.time({
    out_te <- initiators(dat,
                         id = "ID",
                         period = "t",
                         treatment = "A",
                         outcome = "Y",
                         eligible = "eligible",
                         estimand_type = "ITT",
                         model_var = "assigned_treatment",
                         outcome_cov = c("X2", "X4"),
                         use_censor_weights = TRUE,
                         cense = "C",
                         pool_cense = "both",
                         cense_n_cov = ~ period + I(period^2),
                         cense_d_cov = ~ X2 + X4 + period + I(period^2),
                         include_followup_time = ~ followup_time + I(followup_time^2))
  })

  # stop profiling memory
  Rprofmem(NULL)

  # read memory alloc and save time and memory
  mem_lines <- readLines(mem_file)
  alloc_bytes <- sum(as.numeric(str_extract(mem_lines, "\\d+")), na.rm = TRUE)
  alloc_kb <- alloc_bytes / 1024  # convert to kilobytes

  results <- add_row(results,
                     file = basename(f),
                     time_seconds = t["elapsed"],
                     alloc_kb = alloc_kb)
}

# Save results
write_csv(results, "02_appendix_F/out/performance_summary_noCI_r.csv")

results
